### PR TITLE
Deprecating + and - for SignedInteger and its Stride

### PR DESCRIPTION
**Pull request by moiseev**
It leads to mixed-type arithmetics that is not supposed to work.
It was needed before the new collection indexing model to make moving
indexes simple.

### CODE REVIEW

**Comment by moiseev** [comment]
@swift-ci Please test

File: stdlib/public/core/Stride.swift.gyb [hunk 1]
```diff
@@ -62,30 +62,53 @@ public func == <T : Strideable>(x: T, y: T) -> Bool {
   return x.distance(to: y) == 0
 }
 
-public func + <T : Strideable>(lhs: T, rhs: T.Stride) -> T {
+% for Base in ['Strideable', 'SignedInteger']:
+%   deprecated = Base == 'SignedInteger'
+
+%   if deprecated:
+@available(swift, deprecated: 3.1, obsoleted: 4.0, message: "Mixtype addition is deprecated. Please use explicit type conversion.")
```
Comment (jrose-apple) [inline stdlib/public/core/Stride.swift.gyb]: Can you put "Mixed-type" instead of "Mixtype"? I don't think the latter is a standard term.

**Comment by moiseev** [comment]
@jrose-apple fixed [here](https://github.com/apple/swift/pull/6603/commits/457b4a53270da62f1696422b5ed1536312b14995).

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please test Linux platform

**Comment by moiseev** [comment]
@swift-ci Please test Linux platform

**Comment by moiseev** [comment]
@swift-ci Please test macOS platform

**Comment by karwa** [comment]
This seems weird - so every strideable _except_ SignedInteger can actually be advanced by its Stride? If you can only stride by the same type, surely Stride should be equal to Self for the integer types?

It kind of makes our model for Strideable and integers hard to follow. On the one hand, we say Stride != Self for integers, because that's not safe, then we say you can't + with a Stride (and should instead use a Self), because that's not safe either.

**Comment by moiseev** [comment]
@karwa Pleas note this is only the `+` and `-` operators. Integer types still conform to `Strideable` and one can always call `advanced(by:)` and `distance(to:)` methods.

**Comment by moiseev** [comment]
For example the code in the test that's failing on Linux is `let x = puts(s) + 32`. From the looks of it, it is correct, but can you tell the type of the result? One would think that since `+` is only homogeneous, the result type would be whatever `puts` returns, which is `Int32`, but because of the existence of the `+` overload for `Strideable`, `32` gets type `Int` and `x` becomes an `Int` as well.

**Comment by karwa** [comment]
It's still weird - if it's a safe operation, why disallow it?

As for the test, if SignedInteger inherits Strideable, any overloads using SignedInteger should be considered more specialised than the generic Strideable one, no?

**Comment by moiseev** [comment]
> It's still weird - if it's a safe operation, why disallow it?
Because Swift does not allow mixed-type arithmetics. And this one was a violation, no longer necessary because of the changed collection indexing model.

> As for the test, if SignedInteger inherits Strideable, any overloads using SignedInteger should be considered more specialised than the generic Strideable one, no?
Are you talking about the `+` that's defined for `(Int32, Int32)`? Unfortunately, the first type that's tried for the literal `32` is Int, and the compiler is able to typecheck the expression using `(Int32, Int)` overload, so the `Int32` is never even tried on the literal.

**Comment by karwa** [comment]
The thing is that I use SignedInteger, and Range<T: SignedInteger>, and I'm afraid my code is going to become harder to read with this.

> Are you talking about the + that's defined for (Int32, Int32)? Unfortunately, the first type that's tried for the literal 32 is Int, and the compiler is able to typecheck the expression using (Int32, Int) overload, so the Int32 is never even tried on the literal.

Maybe we can teach the compiler that, when resolving operators on integer literals, it should prefer same-type operators? Literal resolution is a special-case; nobody can define their own types of literals (like color literals) without compiler support anyway. So if we want to improve literal resolution without breaking strongly-typed Range operations, maybe it's better to teach the compiler that `(SignedInteger, SignedInteger.Stride)` isn't always the preferred solution for operators.

Or we change SignedInteger.Stride to equal Self, then both cases just work.

**Comment by moiseev** [comment]
@karwa can you provide an example of the code please?
/cc @DougGregor to comment on the compiler side of things.

`SignedInteger.Stride` cannot be `Self`, because distance between `Int8.min` and `Int8.max` is not representable in `Int8`. Same argument, obviously, applies to `Int` being the `Stride` for `Int`, but after a lot of discussion we decided `Int` to be sufficient for most practical purposes. "640k is ought to be enough for everyone." (C)

**Comment by karwa** [comment]
One of the types I use is a lot is a "RangeCollection", basically a RLE array of some tag values. Keeping the constraints on `Bound` broad is really helpful; it means that I can tag ranges of collection indexes even when those indexes are complex. If it's a signed integer though, I can provide easier APIs for things like inserting and removing indexes and better guarantees about what happens at edge-cases.

Example:

```swift

struct RangeCollection<Bound, Tag> where Bound: Comparable {
  typealias TaggedRange = (Range<Bound>, Tag)
  typealias Database    = Array<TaggedRange>

  var dataBase = Database()
}

extension RangeCollection where Bound: Strideable {

  mutating func shiftTags(from: Database.Index, by distance: Bound.Stride, using shift: @escaping (Bound, Bound.Stride)->Bound) 
    dataBase[from..<dataBase.endIndex] = dataBase[from..<dataBase.endIndex].map { tr in
      let newRange = shift(tr.0.lowerBound, distance)..<shift(tr.0.upperBound, distance) as Range
      return (newRange, tr.1)
    }.dropFirst(0)
  }
}

extension RangeCollection where Bound: Strideable, Bound.Stride: SignedInteger {

  mutating func delete(in range: Range<Bound>) {

    // Find the index of the given range within the Database
    // ...
    // Delete the tagged ranges
    // ...
    // Shift all later taggedranges up
    shiftTags(from: dbIndex, by: range.count, using: -) // - will be deprecated; must use...?
  }
}
```

**Comment by moiseev** [comment]
@karwa I don't think this particular example will be affected by the deprecation. Your `-` is of type `(Strideable, Strideable.Stride)` which is not deprecated. Precisely because the `Bound` is just `Strideable` and not `SignedInteger`.

**Comment by moiseev** [comment]
Just tried the snippet with the given change and it does not issue a warning for `extension RangeCollection where Bound : Strideable`.

**Comment by karwa** [comment]
Ok, well in any case I'm concerned that perfectly valid code becomes unavailable because of what is really a compiler limitation. While it's true that we don't support _arbitrary_ heterogeneous arithmetic, arithmetic between a `SignedInteger` and its `Stride` is sort of the entire point of the `Strideable` protocol. Deprecating one syntax in favour of another equivalent one simply masks a deficiency of the compiler.

**Comment by moiseev** [comment]
@swift-ci Please test macOS platform

**Comment by moiseev** [comment]
@swift-ci Please smoke test and merge

**Comment by karwa** [comment]
Any response to making the warning less absolute (e.g. removing the `obsolete` part or suggesting `.advanced(by:)` as an alternative to casting)?

Specifically, I'm concerned that if the same operation is described as being bad in one place and perfectly fine in another, it sends mixed messages regarding numerical safety. "Obsolete" is not really technically accurate, anyway - for one, we could never remove the overload, and secondly, a future version of the compiler may do better at inferring types for integer literals with operators, at which point we could restore it.

**Comment by jrose-apple** [comment]
I'm with Max on this. The operation is *wrong* according to the current semantics of Swift numbers—Strideable notwithstanding—and it *does* lead to programming mistakes, which is why the old _DisallowMixedTypeArithmetic overloads were there. If anything, {{obsolete}} may help the type-checker figure out the appropriate type to infer for the literal case.

**Comment by moiseev** [comment]
@karwa You are right that it is the same operation, essentially, however, when it's called as `advanced(by:)` you are likely in a much more limited context, moving the collection index, for example, whereas when you simply do some arithmetics, it is too easy to overlook the error.

I understand your desire to keep your code clean passing a simple `using: +` into your methods. Defining your own operator for 'strideable plus' is quite straightforward.

FWIW, the new integer protocols are coming eventually, and there will be an `Arithmetic` protocol, that you would probably be able to use.

**Comment by moiseev** [comment]
@swift-ci Please smoke test

**Comment by moiseev** [comment]
@swift-ci Please test macOS platform
